source-firestore/pull.go

```diff
@@ -289,40 +399,157 @@ func (c *capture) Run(ctx context.Context) error {
+
+	var cursor *firestore.DocumentSnapshot
+	if resumeState == nil {
+		// This should never happen since we only run BackfillAsync when there's a backfill to perform, but seemed safe enough to check anyway
+		return nil
+	} else if resumeState.Cursor == "" {
+		// If the cursor path is empty then we just leave the cursor document pointer nil
+		logEntry.Info("starting async backfill")
+	} else if resumeDocument, err := client.Doc(resumeState.Cursor).Get(ctx); err != nil {
+		// If we fail to fetch the resume document, we clear the relevant backfill cursor and error out.
+		// This will cause the backfill to restart from the beginning after the capture gets restarted,
+		// and in the meantime it will show up as an error in the UI in case there's a persistent issue.
+		if checkpointJSON, err := c.State.UpdateBackfillState(collectionID, &backfillState{}); err != nil {
+			return err
+		} else if err := c.Output.Checkpoint(checkpointJSON, true); err != nil {
+			return err
+		}
+		return fmt.Errorf("restarting backfill %q: error fetching resume document %q", collectionID, resumeState.Cursor)
+	} else if !resumeDocument.UpdateTime.Equal(resumeState.MTime) {
+		// Just like if the resume document fetch fails, mtime mismatches cause us to error out, so
+		// we'll restart from the beginning when the connector gets restarted and in the meantime
+		// it'll show up red in the UI.
+		if checkpointJSON, err := c.State.UpdateBackfillState(collectionID, &backfillState{}); err != nil {
+			return err
+		} else if err := c.Output.Checkpoint(checkpointJSON, true); err != nil {
+			return err
+		}
+		return fmt.Errorf("restarting backfill %q: resume document %q modified during backfill", collectionID, resumeState.Cursor)
+	} else {
+		cursor = resumeDocument
+	}
+
+	var numDocuments int
+	for {
+		var query firestore.Query = client.CollectionGroup(collectionID).Query
+		if cursor != nil {
+			query = query.StartAfter(cursor)
+		}
+		query = query.Limit(backfillChunkSize)
+
+		var docs, err = query.Documents(ctx).GetAll()
+		if err != nil {
+			if status.Code(err) == codes.Canceled {
+				err = context.Canceled // Undo an awful bit of wrapping which breaks errors.Is()
+			}
+			return fmt.Errorf("error backfilling %q: chunk query failed after %d documents: %w", collectionID, numDocuments, err)
+		}
+		logEntry.WithFields(log.Fields{
+			"total": numDocuments,
+			"chunk": len(docs),
+		}).Debug("processing backfill documents")
+		if len(docs) == 0 {
+			break
+		}
+
+		for _, doc := range docs {
+			logEntry.WithField("doc", doc.Ref.Path).Trace("got document")
+
+			// We update the cursor before checking whether this document is being
+			// backfilled. This does, unfortunately, mean that it's possible for changes
+			// to a document which *isn't being captured* could break the ongoing backfill
+			// resume behavior, but that's just how Firestore collection group queries
+			// work.
+			cursor = doc
```

Comment:
The cursor is used for every chunk query of 4,096 documents, so if we don't update it to a non-captured document at all and there are more than 4,096 such documents in a row we'll just spin forever and not make forward progress. But I think I see what you're getting at, we could decouple the cursor used during a backfill a bit from what gets persisted to the state checkpoint, so the persisted resume cursor will only ever be a captured document.

However this would introduce a potential failure mode where the collection group queries start with a massive dataset that isn't supposed to be captured, we start working our way through that, and then get restarted repeatedly and basically we keep causing a ton of read traffic without actually making forward progress (until the connector is left running long enough that it can complete all of that and reach a captured document and update the state checkpoint to that). I'm not sure if this is preferable, but either way I think we can adjust the behavior in the future if it proves necessary.

Reply:
That sounds reasonable, I wanted to make sure I understand the implication of this 👍🏽